diff --git a/third_party/lcms/src/cmspack.c b/third_party/lcms/src/cmspack.c
index 3982cc564..7f55ce304 100644
--- a/third_party/lcms/src/cmspack.c
+++ b/third_party/lcms/src/cmspack.c
@@ -3864,9 +3864,6 @@ cmsUInt32Number CMSEXPORT cmsFormatterForPCSOfProfile(cmsHPROFILE hProfile, cmsU
     cmsUInt32Number nOutputChans = cmsChannelsOf(ColorSpace);
     cmsUInt32Number Float = lIsFloat ? 1U : 0;
 
-    // Unsupported color space?
-    if (nOutputChans < 0) return 0;
-
     // Create a fake formatter for result
     return FLOAT_SH(Float) | COLORSPACE_SH(ColorSpaceBits) | BYTES_SH(nBytes) | CHANNELS_SH(nOutputChans);
 }
