diff --git a/third_party/libopenjpeg/opj_includes.h b/third_party/libopenjpeg/opj_includes.h
index 13613ce52..34235921f 100644
--- a/third_party/libopenjpeg/opj_includes.h
+++ b/third_party/libopenjpeg/opj_includes.h
@@ -220,7 +220,6 @@ typedef unsigned int OPJ_BITFIELD;
 
 #define OPJ_UNUSED(x) (void)x
 
-#include "opj_clock.h"
 #include "opj_malloc.h"
 #include "event.h"
 #include "function_list.h"
